Issue4413 caller argument expression

manfred-brands opened the pull request:
@OsirisTerje I see a problem, the compiler cannot distinguish between:

```csharp
Assert.That(actual, constraint, [CAE] string actualExpression =  "", [CAE] string constraintExpression =  "")
{}
Assert.That(actual, constraint, string? message, [CAE] string actualExpression =  "", [CAE] string constraintExpression = "")
{}
```
When calling it like:
```
Assert.That(actual, constraint, message)
```

Because it matches both.

manfred-brands commented:
See here:
![image](https://github.com/nunit/nunit/assets/8655218/630af383-3452-44e0-b466-01d1c3fbe17b)

It thinks we passed an `actualMessage` not a `message`.

OsirisTerje commented:
Got the same.  I changed the nullable message strings into non-null.  That sølvet it.   Suggest you merge my cae branch into your.  I can merge back again tomorrow to get your changes in.

OsirisTerje commented:
@manfred-brands  Partially merged this into the CallerArgumentExpression branch, PR #4427 

We should handle conditional formatting as an addition afterwards.  I am not sure why we actually need it btw, but perhaps some discussion I have missed?

There are no overloads that don't compile now.

manfred-brands commented:
> We should handle conditional formatting as an addition afterwards. I am not sure why we actually need it btw, but perhaps some discussion I have missed?

String formatting is a very expensive operation. Doing this for every unit tests will make tests run much slower. I'll see if I can create a benchmark test quantifying this.

manfred-brands commented:
@OsirisTerje Once I noticed that you removed the overloads with no _message_ parameter, I could finish mine.
Seeing what you did, I fixed the nunit test asserting message contains from 'Is.EqualTo' into 'Does.Contain'.
I added nunit tests to assert that the message contains the _actual_ and _constraint_ expressions.

Note that there is no conditional on nunit462/net6.0. the CAE works for all TargetFrameworks.

OsirisTerje commented on src/NUnitFramework/tests/Attributes/LifeCycleAttributeTests.cs:
```diff
@@ -53,7 +53,7 @@ public void InstancePerTestCaseFullLifeCycleTest()
 
             ITestResult result = TestBuilder.RunTest(fixture);
-            Assert.That(result.ResultState.Status, Is.EqualTo(TestStatus.Passed), result.Message);
+            Assert.That(result.ResultState.Status, Is.EqualTo(TestStatus.Passed), (NUnitString)result.Message);
```
Is the typecast needed?

manfred-brands replied:
Not anymore. Removed.

manfred-brands commented on src/NUnitFramework/framework/Assert.ThatAsync.cs:
```diff
@@ -29,17 +45,22 @@ public static Task ThatAsync(AsyncTestDelegate code, IResolveConstraint constrai
-        public static async Task ThatAsync(AsyncTestDelegate code, IResolveConstraint constraint, string? message)
+        public static async Task ThatAsync(AsyncTestDelegate code, IResolveConstraint constraint,
+            FormattableString message,
+#pragma warning disable CS1573 // Parameter has no matching param tag in the XML comment (but other parameters do)
```
I do agree about the clutter. I had initially thought of only doing one at the top of file as well.
The reason I didn't is that we would then miss errors if a real field was added to a method without corresponding documentation. If you think that won't be an issue I can reduce to a single disable.

manfred-brands replied:
Changed this in a separate commit. Depending on your answer above we can remove or keep it.

manfred-brands commented on src/NUnitFramework/framework/Properties/AssemblyInfo.cs:
```diff
@@ -40,6 +40,12 @@
+#if WINDOWS
+[assembly: AssemblyTitle("NUnit Framework (.NET 6.0)")]
+#else
+[assembly: AssemblyTitle("NUnit Framework (.NET 6.0-windows)")]
```
Indeed forgotten to remove. Also removed NETSTANDARD2_0 and more from CommonAssembly.cs

manfred-brands commented on src/NUnitFramework/framework/Assert.cs:
```diff
@@ -306,12 +305,13 @@ public static async Task MultipleAsync(AsyncTestDelegate testDelegate)
         #region Helper Methods
-        private static void ReportFailure(ConstraintResult result, string? message)
+
+        internal static string? ExtendedMessage(string methodName, string? message, string actualExpression, string constraintExpression)
```
The return type indeed not.  Changed.

The _message_ comes from the _getExceptionMessage` overloads which are declared as `Func<string?>`.
Also there are some NUnit tests (e.g. `LifeCycleAttributeTest`) that pass in `result.Message` which is also declared `string?`.
If we want to change all that, I would do that in a separate PR.

manfred-brands commented on src/NUnitFramework/framework/NUnitString.cs:
```diff
@@ -0,0 +1,48 @@
+        /// <param name="message">The message formattable to hold.</param>
+        public NUnitString(string? message)
+        {
+            _message = message;
```
We can change it, depending on the outcome above.
If `null` is still a valid parameter, it is better to leave it as that then add an extra conditional.
The place where it is relevant (`Assert.ExtendedMessage`) uses `string.IsNullOrEmpty` so can deal with either.

OsirisTerje replied:
Sounds good.  
The more nulls we can fix up, the better.
I added that as a task in #4431

manfred-brands commented on src/NUnitFramework/framework/Assert.ThatAsync.cs:
```diff
@@ -16,10 +17,25 @@ public abstract partial class Assert
+        public static async Task ThatAsync(AsyncTestDelegate code, IResolveConstraint constraint,
+            NUnitString message = default,
+#pragma warning disable CS1573 // Parameter has no matching param tag in the XML comment (but other parameters do)
+            [CallerArgumentExpression(nameof(code))] string actualExpression = "",
```
I didn't get the warnings. Although intellisense didn't want to help me here.
The language version is dependent on the compiler, not the runtime. As we use the NET7.0 SDK, we can use LangVersion11.

manfred-brands replied:
Actually setting the LangVersion to 11 causes the intellisense to work.
